### ISSUE TITLE
Unification and documentation of internal representation of hyperparameters

### ISSUE THREAD

**Issue by mfeurer** [automl/ConfigSpace#252]
TLDR: All hyperparameters have an internal float representation (yes, even categorical ones), which is either in the unit hypercube, or which is not in the unit hypercube, and either distributed uniformly in the unit hypercube, or not. This is hard to understand for anyone using or contributing to the ConfigSpace and should be unified.

Longer:

All hyperparameters have an internal float representation to

1. simplify the usage of the ConfigSpace package for global optimization packages such as SMAC, and
2. optimize for speed with Cython by having everything in numpy arrays.

The internal float representation is [0, ..., n_members] for categorical hyperparameters and ordinal hyperparameters. However, for continuous hyperparameters it is not uniform. The following hyperparameters scale into the unit hypercube:

* UniformFloatHyperparameter
* UniformIntegerHyperparameter

The following ones do not:

* NormalFloatHyperparameter
* NormalIntegerHyperparameter

The following ones scale into the unit hypercube, but are not distributed uniformly (please read [here](https://github.com/automl/ConfigSpace/pull/222#issuecomment-1058015316) for a discussion between @hvarfner and me about this topic).

* BetaFloatHyperparameter
* BetaIntegerHyperparameter

I suggest to do three steps:
1. Decide on a standard way of implementing everything. Ideally, we can reuse a lot of functions (sampling random hyperparameters, neighborhood generation, converting between the uniform and the true distribution) from a common base class. I'd suggest the following:
    1. Every numerical hyperparameter lives in a uniformly distributed integer hypercube. This indeed warps the space, but I think, given the user provided prior information on the space, this warping should be done (and we indeed do it for log-hyperparameters).
    2. When sampling, we sample uniformly between 0 and 1 and use the function `distribution.cdf` and `distribution.ppf` to transform between the two representations.
    3. Similarly, for generating neighbors, we walk in the uniformly distributed unit hypercube and use the functions mentioned above to transform into the real space.
4. Implement this
5. Document this (see also https://github.com/automl/ConfigSpace/issues/239)

With this we can then also properly implement a truncated normal distribution, which is right now [hacked into the normal distribution](https://github.com/automl/ConfigSpace/issues/200).

CC @hvarfner (b/c PiBO) @mlindauer (b/c SMAC) @KEggensperger (b/c SMAC) @dengdifan (b/c SMAC) @frank-hutter (b/c piBO) - please add others who should be interested in this topic to this thread if I forgot them
